File: experiments/problem1_nonperiodicity/analyze_period_two_phase_frontier_projection.cpp
#include <algorithm>
#include <array>
#include <cstdint>
#include <iomanip>
#include <iostream>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

namespace {
constexpr int MAXIMUM_COMPLEXITY = 25;
constexpr std::uint64_t KNOWN_COUNTEREXAMPLE = 0x1bcd3a7b3fdfbULL;
constexpr char KNOWN_WORD[] = "uuuttttutuptttututututuu";

std::uint64_t forward_generator(char name, std::uint64_t state) {
    const std::uint64_t stepped = state ^ ((state << 1) | (state << 2));
    if (name == 't') return stepped;
    if (name == 'u') return stepped ^ 1;
    if (name == 'p') return stepped ^ 1 ^ ((state & 1) == 0 ? 2 : 0);
    throw std::invalid_argument("unknown generator");
}

std::uint64_t projected_child_formula(std::uint64_t parent) {
    const std::uint64_t quotient = parent >> 2;
    switch (parent & 3) {
        case 0: return forward_generator('t', quotient);
        case 1: return forward_generator('u', quotient);
        case 2:
        case 3: return forward_generator('p', quotient);
    }
    throw std::runtime_error("unreachable residue");
}

int bit_length(std::uint64_t value) {
    return value == 0 ? 0 : 64 - __builtin_clzll(value);
}

std::vector<std::uint64_t> next_level(const std::vector<std::uint64_t>& states) {
    std::vector<std::uint64_t> next;
    next.reserve(states.size() * 3);
    for (std::uint64_t state : states) {
        const std::uint64_t stepped = forward_generator('t', state);
        next.push_back(stepped);
        next.push_back(stepped ^ 1);
        if ((state & 1) == 0) next.push_back(stepped ^ 3);
    }
    std::sort(next.begin(), next.end());
    next.erase(std::unique(next.begin(), next.end()), next.end());
    return next;
}

bool contains(const std::vector<std::uint64_t>& states, std::uint64_t value) {
    return std::binary_search(states.begin(), states.end(), value);
}

struct LevelSummary {
    std::uint64_t outputs = 0;
    std::uint64_t projected = 0;
    std::uint64_t uncovered = 0;
    std::array<std::uint64_t, 5> fibers{};
};

LevelSummary projection_summary(
    const std::vector<std::uint64_t>& previous,
    const std::vector<std::uint64_t>& current
) {
    LevelSummary summary;
    summary.outputs = current.size();
    std::size_t previous_index = 0;
    std::size_t current_index = 0;
    while (current_index < current.size()) {
        const std::uint64_t parent = current[current_index] >> 2;
        std::size_t end = current_index + 1;
        while (end < current.size() && (current[end] >> 2) == parent) ++end;
        const std::size_t fiber = end - current_index;
        if (fiber < 2 || fiber > 4) {
            throw std::runtime_error("projection fiber outside 2..4");
        }
        while (previous_index < previous.size() && previous[previous_index] < parent) {
            ++previous_index;
        }
        if (previous_index == previous.size() || previous[previous_index] != parent) {
            throw std::runtime_error("projection left preceding frontier");
        }
        ++summary.projected;
        ++summary.fibers[fiber];
        current_index = end;
    }
    summary.uncovered = previous.size() - summary.projected;
    return summary;
}

std::uint64_t apply_word(std::uint64_t state, const std::string& word) {
    for (char name : word) state = forward_generator(name, state);
    return state;
}

void verify_nonconverse(
    char phase,
    int complexity,
    const std::vector<std::uint64_t>& parent_level,
    const std::vector<std::uint64_t>& child_level
) {
    if (phase == 'p' && complexity == 3) {
        if (!contains(parent_level, 12)
            || !contains(child_level, 50)
            || !contains(child_level, 51)
            || contains(child_level, 48)
            || contains(child_level, 49)) {
            throw std::runtime_error("phase-p strict nonconverse example failed");
        }
    }
    if (phase == 'u' && complexity == 4) {
        if (!contains(parent_level, 26)
            || !contains(child_level, 104)
            || !contains(child_level, 105)
            || !contains(child_level, 107)
            || contains(child_level, 106)) {
            throw std::runtime_error("phase-u strict nonconverse example failed");
        }
    }
}

struct PhaseTotals {
    std::uint64_t outputs_checked = 0;
    std::uint64_t formula_checks = 0;
    std::uint64_t projected_parents = 0;
    std::uint64_t uncovered_parents = 0;
    std::array<std::uint64_t, 5> fibers{};
    std::uint64_t checksum = 1469598103934665603ULL;
};

PhaseTotals run_phase(char phase) {
    std::vector<std::uint64_t> states{
        phase == 'p' ? std::uint64_t{3} : std::uint64_t{1}
    };
    PhaseTotals totals;
    std::cout << "phase=" << phase << '\n';
    for (int complexity = 1; complexity <= MAXIMUM_COMPLEXITY; ++complexity) {
        const int expected_bits = phase == 'p'
            ? 2 * complexity
            : 2 * complexity - 1;
        for (std::uint64_t state : states) {
            if (bit_length(state) != expected_bits) {
                throw std::runtime_error("frontier bit-length law failed");
            }
            const std::uint64_t expected = projected_child_formula(state);
            for (char name : {'t', 'u', 'p'}) {
                if ((forward_generator(name, state) >> 2) != expected) {
                    throw std::runtime_error("projection identity failed");
                }
                ++totals.formula_checks;
            }
            totals.checksum ^= state + static_cast<std::uint64_t>(complexity);
            totals.checksum *= 1099511628211ULL;
        }
        totals.outputs_checked += states.size();

        if (complexity == MAXIMUM_COMPLEXITY) {
            std::cout << "level=" << complexity
                      << " distinct_outputs=" << states.size()
                      << '\n';
            break;
        }

        std::vector<std::uint64_t> next = next_level(states);
        const LevelSummary summary = projection_summary(states, next);
        verify_nonconverse(phase, complexity + 1, states, next);
        totals.projected_parents += summary.projected;
        totals.uncovered_parents += summary.uncovered;
        for (int size = 2; size <= 4; ++size) {
            totals.fibers[size] += summary.fibers[size];
        }
        std::cout << "level=" << (complexity + 1)
                  << " distinct_outputs=" << summary.outputs
                  << " projected_parents=" << summary.projected
                  << " preceding_outputs=" << states.size()
                  << " uncovered_preceding=" << summary.uncovered
                  << " fibers_2=" << summary.fibers[2]
                  << " fibers_3=" << summary.fibers[3]
                  << " fibers_4=" << summary.fibers[4]
                  << '\n';
        states.swap(next);
    }
    std::cout << "phase_outputs_checked=" << totals.outputs_checked << '\n';
    std::cout << "phase_formula_checks=" << totals.formula_checks << '\n';
    std::cout << "phase_projected_parents=" << totals.projected_parents << '\n';
    std::cout << "phase_uncovered_parents=" << totals.uncovered_parents << '\n';
    std::cout << "phase_fibers_2=" << totals.fibers[2] << '\n';
    std::cout << "phase_fibers_3=" << totals.fibers[3] << '\n';
    std::cout << "phase_fibers_4=" << totals.fibers[4] << '\n';
    std::cout << "phase_checksum=0x" << std::hex << std::setw(16)
              << std::setfill('0') << totals.checksum << std::dec
              << std::setfill(' ') << '\n';
    return totals;
}
}  // namespace

int main() {
    const std::uint64_t reproduced = apply_word(1, KNOWN_WORD);
    if (reproduced != KNOWN_COUNTEREXAMPLE) {
        throw std::runtime_error("known counterexample generator word failed");
    }

    const PhaseTotals p = run_phase('p');
    const PhaseTotals u = run_phase('u');

    const std::array<int, 3> depths{12, 14, 16};
    std::cout << "known_counterexample=0x" << std::hex
              << KNOWN_COUNTEREXAMPLE << std::dec << '\n';
    for (int depth : depths) {
        const std::uint64_t ancestor = KNOWN_COUNTEREXAMPLE >> (2 * depth);
        const int residual_complexity = 25 - depth;
        const int expected_bits = 2 * residual_complexity - 1;
        if (bit_length(ancestor) != expected_bits) {
            throw std::runtime_error("known ancestor bit-length mismatch");
        }
        std::cout << "known_ancestor depth=" << depth
                  << " residual_complexity=" << residual_complexity
                  << " value=0x" << std::hex << ancestor << std::dec
                  << " bit_length=" << bit_length(ancestor)
                  << '\n';
    }

    std::cout << "total_outputs_checked="
              << p.outputs_checked + u.outputs_checked << '\n';
    std::cout << "total_formula_checks="
              << p.formula_checks + u.formula_checks << '\n';
    std::cout << "projection_theorem=true\n";
    std::cout << "nonempty_fiber_size_range=2..4\n";
    return 0;
}

File: experiments/problem1_nonperiodicity/analyze_period_two_phase_frontier_projection.py
#!/usr/bin/env python3
"""Analyze the exact base-four projection tower of ordinary phase frontiers.

For phase a in {p,u}, let O_(a,k) be the distinct ordinary outputs at
complexity k.  The projection pi(x)=x>>2 sends every O_(a,k) state into
O_(a,k-1).  Iterating this theorem couples every phase-minimizer witness to an
exact residual-complexity high ancestor.

This is a structural theorem about ordinary phase frontiers.  It does not prove
that the actual moving-fringe orbit has positive return penalties.
"""
from __future__ import annotations

import argparse
import hashlib
import json
from collections import Counter
from typing import Any

PHASES = ("p", "u")
DEFAULT_MAXIMUM_COMPLEXITY = 16
ABSOLUTE_MAXIMUM_COMPLEXITY = 20
KNOWN_COUNTEREXAMPLE = 0x1BCD3A7B3FDFB
KNOWN_GENERATOR_WORD = "uuuttttutuptttututututuu"
KNOWN_DEPTHS = (12, 14, 16)


class ProjectionLimitError(RuntimeError):
    """Raised before a controlled frontier campaign exceeds its cap."""


def forward_generator(name: str, state: int) -> int:
    if state < 0:
        raise ValueError("state must be nonnegative")
    stepped = state ^ ((state << 1) | (state << 2))
    if name == "t":
        return stepped
    if name == "u":
        return stepped ^ 1
    if name == "p":
        return stepped ^ 1 ^ (2 if state & 1 == 0 else 0)
    raise ValueError("unknown generator")


def frontier_children(state: int) -> tuple[int, ...]:
    stepped = forward_generator("t", state)
    if state & 1:
        return stepped, stepped ^ 1
    return stepped, stepped ^ 1, stepped ^ 3


def phase_start(phase: str) -> int:
    if phase == "p":
        return 3
    if phase == "u":
        return 1
    raise ValueError("phase must be p or u")


def projected_child_name(parent_residue_mod_4: int) -> str:
    if parent_residue_mod_4 == 0:
        return "t"
    if parent_residue_mod_4 == 1:
        return "u"
    if parent_residue_mod_4 in (2, 3):
        return "p"
    raise ValueError("residue must be in 0..3")


def projected_child_formula(parent: int) -> int:
    if parent < 0:
        raise ValueError("parent must be nonnegative")
    quotient, residue = divmod(parent, 4)
    return forward_generator(projected_child_name(residue), quotient)


def verify_projection_identity(parent: int) -> None:
    expected = projected_child_formula(parent)
    for name in ("t", "u", "p"):
        observed = forward_generator(name, parent) >> 2
        if observed != expected:
            raise AssertionError("generator projection identity failed")


def generate_levels(phase: str, maximum_complexity: int) -> list[set[int]]:
    if maximum_complexity < 1:
        raise ValueError("maximum_complexity must be positive")
    states = {phase_start(phase)}
    levels: list[set[int]] = []
    for _ in range(maximum_complexity):
        levels.append(states)
        states = {child for state in states for child in frontier_children(state)}
    return levels


def level_projection_row(
    previous: set[int], current: set[int], complexity: int
) -> dict[str, Any]:
    fibers = Counter(value >> 2 for value in current)
    violations = sorted(parent for parent in fibers if parent not in previous)
    if violations:
        raise AssertionError("phase frontier projection left the preceding level")
    fiber_histogram = Counter(fibers.values())
    if any(size < 2 or size > 4 for size in fibers.values()):
        raise AssertionError("nonempty projection fiber must contain two to four lifts")
    return {
        "complexity": complexity,
        "distinct_outputs": len(current),
        "preceding_outputs": len(previous),
        "projected_parents": len(fibers),
        "uncovered_preceding_outputs": len(previous) - len(fibers),
        "fiber_histogram": {
            str(size): fiber_histogram.get(size, 0) for size in (2, 3, 4)
        },
        "projection_violations": 0,
    }


def verify_iterated_projection(levels: list[set[int]]) -> int:
    checks = 0
    for complexity, states in enumerate(levels, start=1):
        for value in states:
            projected = value
            for depth in range(1, complexity):
                projected >>= 2
                if projected not in levels[complexity - depth - 1]:
                    raise AssertionError("iterated projection theorem failed")
                checks += 1
    return checks


def apply_generator_word(start: int, word: str) -> int:
    state = start
    for name in word:
        state = forward_generator(name, state)
    return state


def known_counterexample_ancestry(u_levels: list[set[int]]) -> list[dict[str, Any]]:
    state = apply_generator_word(phase_start("u"), KNOWN_GENERATOR_WORD)
    if state != KNOWN_COUNTEREXAMPLE:
        raise AssertionError("known generator word did not reproduce counterexample")
    if len(KNOWN_GENERATOR_WORD) + 1 != 25:
        raise AssertionError("known counterexample complexity mismatch")
    rows = []
    for depth in KNOWN_DEPTHS:
        ancestor = state >> (2 * depth)
        residual_complexity = 25 - depth
        if residual_complexity > len(u_levels):
            raise AssertionError("campaign too shallow for known ancestor")
        if ancestor not in u_levels[residual_complexity - 1]:
            raise AssertionError("known projected ancestor left phase-u frontier")
        expected_bits = 2 * residual_complexity - 1
        if ancestor.bit_length() != expected_bits:
            raise AssertionError("known ancestor bit-length law failed")
        rows.append(
            {
                "depth": depth,
                "residual_complexity": residual_complexity,
                "ancestor": ancestor,
                "ancestor_hex": hex(ancestor),
                "bit_length": ancestor.bit_length(),
            }
        )
    return rows


def verify_strict_nonconverse(levels: dict[str, list[set[int]]]) -> list[dict[str, Any]]:
    examples = [
        {
            "phase": "p",
            "parent_complexity": 2,
            "parent": 12,
            "present_lifts": [50, 51],
            "absent_lifts": [48, 49],
        },
        {
            "phase": "u",
            "parent_complexity": 3,
            "parent": 26,
            "present_lifts": [104, 105, 107],
            "absent_lifts": [106],
        },
    ]
    for row in examples:
        phase_levels = levels[row["phase"]]
        parent_level = phase_levels[row["parent_complexity"] - 1]
        child_level = phase_levels[row["parent_complexity"]]
        if row["parent"] not in parent_level:
            raise AssertionError("nonconverse parent missing")
        for value in row["present_lifts"]:
            if value not in child_level or value >> 2 != row["parent"]:
                raise AssertionError("present nonconverse lift failed")
        for value in row["absent_lifts"]:
            if value in child_level or value >> 2 != row["parent"]:
                raise AssertionError("absent nonconverse lift failed")
    return examples


def run_campaign(
    maximum_complexity: int = DEFAULT_MAXIMUM_COMPLEXITY,
) -> dict[str, Any]:
    if not 13 <= maximum_complexity <= ABSOLUTE_MAXIMUM_COMPLEXITY:
        raise ProjectionLimitError(
            "maximum complexity must be between 13 and the controlled cap"
        )
    levels = {
        phase: generate_levels(phase, maximum_complexity) for phase in PHASES
    }
    formula_checks = 0
    phases: dict[str, Any] = {}
    iterated_checks = 0
    for phase in PHASES:
        phase_levels = levels[phase]
        rows = []
        for complexity, states in enumerate(phase_levels, start=1):
            expected_bits = 2 * complexity if phase == "p" else 2 * complexity - 1
            for state in states:
                if state.bit_length() != expected_bits:
                    raise AssertionError("phase frontier bit-length law failed")
                verify_projection_identity(state)
                formula_checks += 3
            if complexity >= 2:
                rows.append(
                    level_projection_row(
                        phase_levels[complexity - 2], states, complexity
                    )
                )
        iterated_checks += verify_iterated_projection(phase_levels)
        phases[phase] = {
            "levels": rows,
            "final_complexity": maximum_complexity,
            "final_distinct_outputs": len(phase_levels[-1]),
        }

    payload: dict[str, Any] = {
        "status": "exact-theorem-and-controlled-exhaustion",
        "maximum_complexity": maximum_complexity,
        "exact_theorems": {
            "one_level_projection": (
                "for every phase a and k>=2, x in O_(a,k) implies "
                "x>>2 in O_(a,k-1)"
            ),
            "iterated_projection": (
                "x in O_(a,k) implies x>>(2L) in O_(a,k-L) "
                "for every 0<=L<k"
            ),
            "fiber_bound": (
                "every nonempty one-level projection fiber contains "
                "between two and four ordinary outputs"
            ),
            "minimizer_coupling": (
                "a complexity-k phase minimizer in a depth-L survivor cylinder "
                "has a residual phase ancestor of complexity k-L in its aligned "
                "high quotient"
            ),
        },
        "projection_identity": {
            "parent_residue_0": "G_g(4r)>>2=t(r)",
            "parent_residue_1": "G_g(4r+1)>>2=u(r)",
            "parent_residue_2": "G_g(4r+2)>>2=p(r)",
            "parent_residue_3": "G_g(4r+3)>>2=p(r)",
            "independent_of_child_generator": True,
        },
        "formula_checks": formula_checks,
        "iterated_membership_checks": iterated_checks,
        "phases": phases,
        "known_counterexample_ancestry": known_counterexample_ancestry(
            levels["u"]
        ),
        "strict_nonconverse_examples": verify_strict_nonconverse(levels),
        "research_consequence": (
            "the aligned high quotient supplies an exact necessary "
            "phase-frontier ancestor for every minimizer witness, but the "
            "ancestor alone is not sufficient to reconstruct a valid low lift"
        ),
        "scientific_boundary": (
            "This theorem couples ordinary phase minimizers to exact high "
            "ancestors. It does not prove recurring positive return penalties, "
            "phase-complexity divergence, exclusion of eventual period two, or "
            "Rule 30 center nonperiodicity."
        ),
    }
    encoded = json.dumps(payload, sort_keys=True, separators=(",", ":")).encode()
    payload["certificate_sha256"] = hashlib.sha256(encoded).hexdigest()
    return payload


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--maximum-complexity",
        type=int,
        default=DEFAULT_MAXIMUM_COMPLEXITY,
    )
    args = parser.parse_args()
    print(
        json.dumps(
            run_campaign(args.maximum_complexity),
            indent=2,
            sort_keys=True,
        )
    )


if __name__ == "__main__":
    main()
